nav2_regulated_pure_pursuit_controller/src/regulated_pure_pursuit_controller.cpp

```diff
@@ -606,20 +601,22 @@ nav_msgs::msg::Path RegulatedPurePursuitController::transformGlobalPlan(
   const double max_costmap_dim = std::max(costmap->getSizeInCellsX(), costmap->getSizeInCellsY());
   const double max_transform_dist = max_costmap_dim * costmap->getResolution() / 2.0;
 
+  auto lookahead_element =
+    nav2_util::geometry_utils::first_element_beyond(
+    global_plan_.poses.begin(), global_plan_.poses.end(), lookahead_dist_);
+
   // First find the closest pose on the path to the robot
   auto transformation_begin =
     nav2_util::geometry_utils::min_by(
-    global_plan_.poses.begin(), global_plan_.poses.end(),
+    global_plan_.poses.begin(), lookahead_element,
     [&robot_pose](const geometry_msgs::msg::PoseStamped & ps) {
       return euclidean_distance(robot_pose, ps);
     });
 
-  // Find points definitely outside of the costmap so we won't transform them.
```

Comment:
I think this was the right implementation, since we want to find the first point that's outside of the statically set costmap size. This isn't an integral distance but a real-constraint distance dependent on the costmap size. This could potentially not prune quite to the end of the costmap (dropping necessary data for longer look-ahead points) for wavy paths. There's no real benefit to culling a few points more off the end of the rolling costmap window.

Reply:
if the transformed_plan is very large and has poses ahead and behind the robot as in overlapping case, the getlookahead point may select the point behind the robot making the robot leave some path and turns back immediately.
for this it is trimmed to a particular distance from transform begin using integrated distance. 

if we need this to be remain unchanged,  we can modify and get the lookahead point using integrated distance.

Reply:
OK that makes sense, but then this value shouldn't be derivative of the costmap size as it currently is e.g.

```
  const double max_transform_dist = max_costmap_dim * costmap->getResolution() / 2.0;
```

This should probably now be a parameterized distance

Reply:
I agree that finding the closest point needs to have an upper bound, so that it doesn't select overlapping poses at a later part of the path. Also noting that we are implicitly relying on the lower bound given by pruning the global plan after receiving it, so we don't process the same segments twice.

Reply:
I agree with @SteveMacenski that we shouldn't prune by lookahead distance: this part of the controller happens before the lookahead distance is computed, right?

I think that we should simply prune by the minimum of the cusp distance and the costmap size. That guarantees that you won't be able to get portions of the path that have looped around.

Reply:
Yes, this happens before any of the lookahead logic is applied. this is just pruning the path forward to only consider things in a local window forward